text/0000-codify-infrastructure.md

```diff
@@ -0,0 +1,392 @@
+
+Currently, the API lives in [streampusher/api](https://github.com/streampusher/api) ([look, here are the model classes](https://github.com/streampusher/api/tree/master/app/models)) and is written in [Ruby](https://www.ruby-lang.org/en/). The API is what's called [RESTful](https://restfulapi.net/), which means that it supports operations like `GET`, which retrieves data, `POST`, which creates new data, `PUT`, which is like `POST` but also updates data, `DELETE`, which is exactly what it sounds like, and `PATCH` which is like `PUT` but semantically different.
+
+REST is pretty good, and certainly better than [SOAP](https://restfulapi.net/soap-vs-rest-apis/), but it has this issue where every particular type of data needs its own endpoint. This means you wind up getting super long and specific URLs like `api.datafruits.fm/artists/ovenrake/podcasts`. It also means that it's hard to change existing endpoints -- like if you wanted to restructure to `api.datafruits.fm/podcasts?dj=ovenrake`, then you'd have to deprecate the old endpoint, update all of your client code, yadda yadda yadda.
+
+Another issue with REST is that it always returns all of the data it has. Even if all you want is the `title` and `link` to a podcast, you'll also get the `tags` and the `image` and the `bandcamp_link` or whatever other fields it has.
+
+The solution to this is the new shiny thing called [GraphQL](https://graphql.org/), which aims to solve the problems that REST has. Refactoring the existing API into GraphQL is gonna probably warrant its own RFC, but I wanted to mention it here so I didn't forget about it.
+
+In terms of containerization, the API literally [already has a Dockerfile](https://github.com/streampusher/api/blob/master/Dockerfile) so this should be ok with only minor tweaks.
+
+#### Audio streaming 🔊
+
+This is probably the second most important part of the ᴅᴀᴛᴀғʀᴜɪᴛs ᴇxᴘᴇʀɪᴇɴᴄᴇ, behind the site itself. The code that handles this stuff lives in [streampusher/radio](https://github.com/streampusher/radio). The audio streaming is built on [icecast](https://icecast.org), which is what broadcasts the audio, and [liquidsoap](https://www.liquidsoap.info/), which provides icecast with the audio to stream.
+
+There's something to be said about rebuilding this setup to be less fragile, and there will likely be some issues trying to move this into Kubernetes. Kubernetes relies on its containers being mostly stateless, and so there may potentially be issues with having containers die in the middle of a stream.
```

Comment:
Can we talk about what is fragile with the current setup? Don't hold back. :)

I feel like the main problem is DJs disconnecting due to them being on wifi instead of ethernet, and that's not exactly something we can fix on our side I think. 

Also sometimes the cert for the relay expires and letsencrypt doesn't seem to renew it automatically, that's definitely something we could fix.

Anything else?

Reply:
> Can we talk about what is fragile with the current setup? Don't hold back. :)

Ahhh lol we'll I'm by *no* means an expert on this stuff. I mostly mean "fragile" in terms of "it's hard to understand what changes will break things." Similarly to how it would be nice to have all of the infrastructure documented, it would be nice to have some more idea about how all the different files in the repo fit together.

> I feel like the main problem is DJs disconnecting due to them being on wifi instead of ethernet, and that's not exactly something we can fix on our side I think.

I agree, and it might be worth tinkering with the [timeout](https://www.liquidsoap.info/doc-dev/harbor.html#parameters).

> Also sometimes the cert for the relay expires and letsencrypt doesn't seem to renew it automatically, that's definitely something we could fix.

Yeah, in our cluster we'd have a dedicated [certificate manager](https://cert-manager.io/docs/) to make sure all our stuff is encrypted, so we'd offload the security functionality to there. This also I think is a benefit because it decouples security from radio.

> Anything else?

Off the top of my head:

- There's three top level files ([start.sh](https://github.com/streampusher/radio/blob/master/start.sh), [setup.sh](https://github.com/streampusher/radio/blob/master/setup.sh), and a [Vagrantfile](https://github.com/streampusher/radio/blob/master/Vagrantfile)) which all seem to serve a similar purpose, except all three of them haven't been touched in years.
- Similarly, there's a *bunch* of files in [/liquidsoap](https://github.com/streampusher/radio/tree/master/liquidsoap) that haven't been touched in years. Some Ruby, some Bash, some config, etc. I'd have a lot more confidence if I could look at the files and know that they're untouched because they're perfect, and not because they're forgotten.
- More generally, I think it makes sense to set up a docker-compose.yml or something in the top-level directory indicating how the different images work together. Unfortunately Kubernetes doesn't play *super* well with docker-compose but there's tools like [Kompose](https://kubernetes.io/docs/tasks/configure-pod-container/translate-compose-kubernetes) that we could drop into CI to auto-magically update the Kubernetes config while also allowing redistribution of the standalone radio stuff.

Reply:
> There's three top level files (start.sh, setup.sh, and a Vagrantfile) which all seem to serve a similar purpose, except all three of them haven't been touched in years.
Similarly, there's a bunch of files in /liquidsoap that haven't been touched in years. Some Ruby, some Bash, some config, etc. I'd have a lot more confidence if I could look at the files and know that they're untouched because they're perfect, and not because they're forgotten.

Sorry about all that, totally my fault. I will try to clean them up a bit.

Reply:
> More generally, I think it makes sense to set up a docker-compose.yml or something in the top-level directory

There is one in [streampusher/api](https://github.com/streampusher/api/blob/master/docker-compose.yml) , were you thinking its not enough?

Reply:
I appreciate it! And no worries, I totally understand getting it up and running and going "ok done with this forever" lol